### PR TITLE
Use concise JSON serialization for FrozenCircuit.

### LINKED ISSUE #3633: Efficient JSON serialization of FrozenCircuit
**Is your feature request related to a use case or problem? Please describe.**

Part of #3235.
Concise serialization (#3601) is available for Cirq objects, but not yet supported for FrozenCircuit.

**Describe the solution you'd like**

FrozenCircuit needs the `_serialization_key_` magic method to enable efficient serialization. This should be a fairly small code change, but may require nontrivial updates to its JSON serialization test.

**What is the urgency from your perspective for this issue? Is it blocking important work?**

P1 - I need this no later than the next release (end of quarter)

### PR DESCRIPTION
**Pull request by 95-martin-orion**
Fixes #3633. This is a rewrite of #3655 that avoids using hash values in serialization; this is possible due to #3673 dynamically applying unique identifiers to each `SerializableByKey` object as it appears.

This PR continues to use FrozenCircuit hashes in circuit diagrams, where hash collision is a less serious issue.

### CODE REVIEW

File: cirq/circuits/frozen_circuit.py
Comment (wcourtney) [inline cirq/circuits/frozen_circuit.py:99]: It doesn't look like this method is used outside of tests. Since it's "private", consider removing it (or using it).

File: cirq/circuits/frozen_circuit.py
Comment (wcourtney) [inline cirq/circuits/frozen_circuit.py:102]: Is this method still intended to be used for serialization? If so, consider a name that indicates this or add a new method that forwards functionality here. Alternatively, maybe a name that declares what this value is rather than how it's used?

File: cirq/circuits/frozen_circuit.py
Comment (wcourtney) [inline cirq/circuits/frozen_circuit.py:97]: IIUC, this poses a risk that a user could add duplicate definitions of circuits to the serialization by providing them with unique names. Maybe it's future work, but should there be a check somewhere to prevent this?

File: cirq/circuits/frozen_circuit.py
Comment (95-martin-orion) [reply cirq/circuits/frozen_circuit.py:99]: This is a "magic method"; defining it here is what enables concise serialization of `FrozenCircuit`s in the JSON serialization protocol.

File: cirq/circuits/frozen_circuit.py
Comment (95-martin-orion) [reply cirq/circuits/frozen_circuit.py:102]: This method is only used in circuit diagrams now - hence the name. Serialization instead uses `_serialization_name_`.

File: cirq/circuits/frozen_circuit.py
Comment (95-martin-orion) [reply cirq/circuits/frozen_circuit.py:97]: We could check during serialization if two `FrozenCircuit`s are equal apart from their names, I guess - unless someone really likes reading JSON-formatted circuits, there probably isn't a use case for identical circuits with different names.

(Aside: thanks to #3673, hash values are not required for concise serialization)

File: cirq/circuits/frozen_circuit.py
Comment (wcourtney) [reply cirq/circuits/frozen_circuit.py:99]: Is it possible to reference the overridden method so that we can easily access its documentation? Should this class also extend [SerializableByKey](https://github.com/quantumlib/Cirq/blob/0677d6082bad22289c8438596dd7bf815c21e1c2/cirq/protocols/json_serialization.py#L332) to be explicit about this support?

File: cirq/circuits/frozen_circuit.py
Comment (wcourtney) [inline cirq/circuits/frozen_circuit.py:104]: Should `self.name or "Circuit"` be `self._serialization_name_()` for consistency, or do we want to keep them independent and they just happen to match today?

**Review by wcourtney — APPROVED** [review]
Mostly nits/optional suggestions at this point, so approving to let you resolve as you see fit.

File: cirq/circuits/frozen_circuit.py
Comment (95-martin-orion) [reply cirq/circuits/frozen_circuit.py:99]: I'm not fundamentally opposed to adding a reference here, but in general magic methods in Cirq don't reference their protocols. Instead, there's an implicit assumption that magic method usage is defined in the relevant protocol; this is supported by the [protocols doc](https://github.com/quantumlib/Cirq/blob/master/docs/protocols.ipynb).

File: cirq/circuits/frozen_circuit.py
Comment (95-martin-orion) [reply cirq/circuits/frozen_circuit.py:104]: It's plausible that we'll want to separate these in the future (e.g. for out-of-line subcircuits in circuit diagrams) - let's keep them independent.

**Comment by 95-martin-orion** [comment]
Thanks Will! It looks like I still need a cirq-maintainers review before merging this.

**Review by mpharrigan — DISMISSED** [review]
1. You're modifying the json test data. The rule is to make new test data files with the new schema but ensure that the old files still load successfully. ~~Unless these classes were never in a released version of Cirq? I can't keep track of the timeline~~ ok `cirq.Circuit` has definitely been around for a while.

2. Adding a name to circuits is a pretty big deal :tm: because it opens the floodgates for users to use it in previously unpredicted ways. I personally think it's a fine idea, but maybe we should shop it around at the cirq cync

File: cirq/protocols/json_test_data/Circuit.json
Comment (mpharrigan) [inline cirq/protocols/json_test_data/Circuit.json:175]: consider omitting this field in the json output when it's not set for fully-backwards-compatible output and generally nicer json

Comment (95-martin-orion) [reply cirq/protocols/json_test_data/Circuit.json]: Done.

**Comment by 95-martin-orion** [comment]
Thanks for catching the backwards-compatibility issue! Should be tested now with the `json_inward` files.

RE circuit names: I don't see much potential for misuse here, given that `Circuit.name` is only used in the serialization of `FrozenCircuit`, but I'm happy to delay until next cynque for confirmation.

**Comment by mpharrigan** [comment]
I know right _now_ it's only used in serialization :)

**Comment by 95-martin-orion** [comment]
Cirq sync discussion: decouple addition of `name` from this PR

**Comment by 95-martin-orion** [comment]
The new, name-free version of this change is #3863.

**Pull request closed without merging**
